core/primitives/src/version.rs

```diff
@@ -61,3 +61,38 @@ impl ProtocolVersionRange {
         self.lower <= version && self.upper.map_or(true, |upper| version < upper)
     }
 }
+
+/// New Protocol features should go here. Features are guarded by their corresponding feature flag.
+/// For example, if we have `ProtocolFeature::EVM` and a corresponding feature flag `evm`, it will look
+/// like
+/// ```
+/// #[cfg(feature = "evm")]
+/// EVM
+/// ```
+#[derive(Hash, PartialEq, Eq)]
+pub enum ProtocolFeature {}
+
+/// Current latest stable version of the protocol.
+#[cfg(not(feature = "nightly_protocol"))]
+pub const PROTOCOL_VERSION: ProtocolVersion = 40;
+
+/// Current latest nightly version of the protocol.
+#[cfg(feature = "nightly_protocol")]
+pub const PROTOCOL_VERSION: ProtocolVersion = 40;
+
+/// Map of feature to the minimal protocol version that introduces the feature. We can determine
+/// whether to apply the new feature by comparing the current protocol version of the network to
+/// `PROTOCOL_FEATURES_TO_VERSION_MAPPING[feature]`.
+#[cfg(not(feature = "nightly_protocol"))]
+lazy_static! {
+    static ref PROTOCOL_FEATURES_TO_VERSION_MAPPING: HashMap<ProtocolFeature, ProtocolVersion> = vec![
+        /* add mapping here */
+    ].into_iter().collect();
+}
+
+#[cfg(feature = "nightly_protocol")]
+lazy_static! {
+    static ref PROTOCOL_FEATURES_TO_VERSION_MAPPING: HashMap<ProtocolFeature, ProtocolVersion> = vec![
+        /* add mapping here */
+    ].into_iter().collect();
+}
```

Comment:
I suggest we bake sanity checks right here (I am not sure if this code works, it is just a draft):

```suggestion
lazy_static! {
    static ref STABLE_PROTOCOL_FEATURES_TO_VERSION_MAPPING: HashMap<ProtocolFeature, ProtocolVersion> = vec![
        /* add mapping here */
    ].into_iter().collect();
}

#[cfg(not(feature = "nightly_protocol"))]
lazy_static! {
    static ref PROTOCOL_FEATURES_TO_VERSION_MAPPING: HashMap<ProtocolFeature, ProtocolVersion> = STABLE_PROTOCOL_FEATURES_TO_VERSION_MAPPING;
}

#[cfg(feature = "nightly_protocol")]
lazy_static! {
    static ref PROTOCOL_FEATURES_TO_VERSION_MAPPING: HashMap<ProtocolFeature, ProtocolVersion> = {
        let nightly_protocol_features_to_version_mapping: HashMap<ProtocolFeature, ProtocolVersion> = vec![
            /* add mapping here */
        ].into_iter().collect();
        for (stable_protocol_feature, stable_protocol_version) in STABLE_PROTOCOL_FEATURES_TO_VERSION_MAPPING.iter() {
            assert!(nightly_protocol_features_to_version_mapping[stable_protocol_feature] >= stable_protocol_version);
        }
        nightly_protocol_features_to_version_mapping
    };
}
```